As per the ClinVar database link provided in the problem, clinvar.vcf.gz file was downloded.
Allele, python library, the vcf file was converted into csv file which can be onened using pandas
#allel.vcf_to_csv('clinvar.vcf', 'example1.csv', fields="*")

Initially we will import essential python libraries for data processisng.

In [1]:
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
#pd.set_option("display.max_rows", None)
import warnings
warnings.filterwarnings('ignore')

In [2]:
#using pandas we are reading the csv file which we got from vcf format and storing it into df 
df = pd.read_csv('example1.csv')

In [3]:
df

CHROM     POS       ID REF             ALT_1  ALT_2  ALT_3  QUAL  \
0          1  861332  1019397   G                 A    NaN    NaN   NaN   
1          1  865519  1125147   C                 T    NaN    NaN   NaN   
2          1  865545  1164676   G                 A    NaN    NaN   NaN   
3          1  865567  1144630   C                 T    NaN    NaN   NaN   
4          1  865568   846933   G                 A    NaN    NaN   NaN   
...      ...     ...      ...  ..               ...    ...    ...   ...   
960101    MT   16011   690273   A                 G    NaN    NaN   NaN   
960102    MT   16013   690274   A                 G    NaN    NaN   NaN   
960103    MT   16015   690276   T  TATTCTCTGTTCTTTC    NaN    NaN   NaN   
960104    MT   16017   690275   T                 C    NaN    NaN   NaN   
960105    MT   16021   690277   C                 T    NaN    NaN   NaN   

        AF_ESP  AF_EXAC  AF_TGP  ALLELEID              CLNDN CLNDNINCL  \
0          NaN      NaN     NaN   1003021       not_provided       NaN   
1          NaN      NaN     NaN   1110865       not_provided       NaN   
2          NaN      NaN     NaN   1153701       not_provided       NaN   
3          NaN      NaN     NaN   1131738       not_provided       NaN   
4          NaN      NaN     NaN    824438       not_provided       NaN   
...        ...      ...     ...       ...                ...       ...   
960101     NaN      NaN     NaN    677665  Juvenile_myopathy       NaN   
960102     NaN      NaN     NaN    677666  Juvenile_myopathy       NaN   
960103     NaN      NaN     NaN    677667  Juvenile_myopathy       NaN   
960104     NaN      NaN     NaN    677668  Juvenile_myopathy       NaN   
960105     NaN      NaN     NaN    677669  Juvenile_myopathy       NaN   

                   CLNDISDB CLNDISDBINCL                       CLNHGVS  \
0           MedGen:CN517202          NaN      NC_000001.10:g.861332G>A   
1           MedGen:CN517202          NaN      NC_000001.10:g.865519C>T   
2           MedGen:CN517202          NaN      NC_000001.10:g.865545G>A   
3           MedGen:CN517202          NaN      NC_000001.10:g.865567C>T   
4           MedGen:CN517202          NaN      NC_000001.10:g.865568G>A   
...                     ...          ...                           ...   
960101  MONDO:MONDO:0010789          NaN        NC_012920.1:m.16011A>G   
960102  MONDO:MONDO:0010789          NaN        NC_012920.1:m.16013A>G   
960103  MONDO:MONDO:0010789          NaN  NC_012920.1:m.16018_16032dup   
960104  MONDO:MONDO:0010789          NaN        NC_012920.1:m.16017T>C   
960105  MONDO:MONDO:0010789          NaN        NC_012920.1:m.16021C>T   

               CLNREVSTAT                  CLNSIG CLNSIGCONF CLNSIGINCL  \
0       criteria_provided  Uncertain_significance        NaN        NaN   
1       criteria_provided           Likely_benign        NaN        NaN   
2       criteria_provided                  Benign        NaN        NaN   
3       criteria_provided           Likely_benign        NaN        NaN   
4       criteria_provided  Uncertain_significance        NaN        NaN   
...                   ...                     ...        ...        ...   
960101  criteria_provided           Likely_benign        NaN        NaN   
960102  criteria_provided  Uncertain_significance        NaN        NaN   
960103  criteria_provided  Uncertain_significance        NaN        NaN   
960104  criteria_provided                  Benign        NaN        NaN   
960105  criteria_provided  Uncertain_significance        NaN        NaN   

                            CLNVC     CLNVCSO CLNVI DBVARID       GENEINFO  \
0       single_nucleotide_variant  SO:0001483   NaN     NaN  SAMD11:148398   
1       single_nucleotide_variant  SO:0001483   NaN     NaN  SAMD11:148398   
2       single_nucleotide_variant  SO:0001483   NaN     NaN  SAMD11:148398   
3       single_nucleotide_variant  SO:0001483   NaN     NaN  SAMD11:148398   
4       single_nucleotide_variant  SO

First we will lok at the unique values present in Chromosome.

In [4]:
df.CHROM.unique()

array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
       20, 21, 22, '22', 'X', 'Y', 'MT'], dtype=object)

The 22nd chromosome is mistakenly written as string 22 which can be replaced by 23, since humans have 23 pairs of chromosome

In [5]:
df['CHROM'].replace({'22': 23}, inplace=True)

In [6]:
#CHecking if the replace method worked
df.CHROM.unique()

array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
       20, 21, 22, 23, 'X', 'Y', 'MT'], dtype=object)

In [7]:
#Finding and creating list of features with nan values
features_with_na = [features for features in df.columns if df[features].isnull().sum()>1]
features_with_na

['ALT_1',
 'ALT_2',
 'ALT_3',
 'QUAL',
 'AF_ESP',
 'AF_EXAC',
 'AF_TGP',
 'CLNDN',
 'CLNDNINCL',
 'CLNDISDB',
 'CLNDISDBINCL',
 'CLNSIG',
 'CLNSIGCONF',
 'CLNSIGINCL',
 'CLNVI',
 'DBVARID',
 'GENEINFO',
 'MC',
 'ORIGIN',
 'RS']

In [8]:
#we will loop through all the features with na value and calculate their nan percentage
for feature in features_with_na:
    print(feature, (np.round(df[feature].isnull().mean(), 4))*100, ' % missing')

ALT_1 0.06  % missing
ALT_2 100.0  % missing
ALT_3 100.0  % missing
QUAL 100.0  % missing
AF_ESP 84.35000000000001  % missing
AF_EXAC 73.81  % missing
AF_TGP 81.98  % missing
CLNDN 0.06  % missing
CLNDNINCL 99.86  % missing
CLNDISDB 0.06  % missing
CLNDISDBINCL 99.86  % missing
CLNSIG 0.06  % missing
CLNSIGCONF 94.86  % missing
CLNSIGINCL 99.86  % missing
CLNVI 76.16000000000001  % missing
DBVARID 99.98  % missing
GENEINFO 0.03  % missing
MC 1.68  % missing
ORIGIN 0.5499999999999999  % missing
RS 34.23  % missing


In [9]:
df.nunique()

CHROM               26
POS             868268
ID              960106
REF              13947
ALT_1             7382
ALT_2                0
ALT_3                0
QUAL                 0
AF_ESP           12923
AF_EXAC          13511
AF_TGP            5420
ALLELEID        960106
CLNDN            29448
CLNDNINCL          422
CLNDISDB          9309
CLNDISDBINCL       397
CLNHGVS         960106
CLNREVSTAT           6
CLNSIG              17
CLNSIGCONF          61
CLNSIGINCL         783
CLNVC                8
CLNVCSO              8
CLNVI            62535
DBVARID            153
GENEINFO         12939
MC                  18
ORIGIN              75
RS              597843
SSR                  1
FILTER_PASS          1
numalt               2
altlen_1          1643
altlen_2             1
altlen_3             1
is_snp               2
dtype: int64

So we are goig to drop all the columns with more than 75% null values and columns with unique values as 1.
we will create a list of all the columns and them drop them using drop function in pandas

In [10]:
# creating a list of all the columns awith null values more thane 80% and unique values as 1
total_col_drop = ['ALT_2', 'ALT_3', 'QUAL', 'AF_ESP', 'AF_EXAC', 'AF_TGP', 'CLNDNINCL', 'CLNDISDBINCL',  'CLNSIGCONF',
               'CLNSIGINCL', 'CLNVI', 'DBVARID', 'altlen_3', 'altlen_2', 'SSR', 'FILTER_PASS', 'CLNDISDB', 'CLNHGVS', 'CLNVCSO', 'GENEINFO', 'RS']

In [11]:
# dropping all the null values and creating a new dataframe as df1
df_1 = df.drop(columns= total_col_drop)

In [12]:
df_1

CHROM     POS       ID REF             ALT_1  ALLELEID  \
0          1  861332  1019397   G                 A   1003021   
1          1  865519  1125147   C                 T   1110865   
2          1  865545  1164676   G                 A   1153701   
3          1  865567  1144630   C                 T   1131738   
4          1  865568   846933   G                 A    824438   
...      ...     ...      ...  ..               ...       ...   
960101    MT   16011   690273   A                 G    677665   
960102    MT   16013   690274   A                 G    677666   
960103    MT   16015   690276   T  TATTCTCTGTTCTTTC    677667   
960104    MT   16017   690275   T                 C    677668   
960105    MT   16021   690277   C                 T    677669   

                    CLNDN         CLNREVSTAT                  CLNSIG  \
0            not_provided  criteria_provided  Uncertain_significance   
1            not_provided  criteria_provided           Likely_benign   
2            not_provided  criteria_provided                  Benign   
3            not_provided  criteria_provided           Likely_benign   
4            not_provided  criteria_provided  Uncertain_significance   
...                   ...                ...                     ...   
960101  Juvenile_myopathy  criteria_provided           Likely_benign   
960102  Juvenile_myopathy  criteria_provided  Uncertain_significance   
960103  Juvenile_myopathy  criteria_provided  Uncertain_significance   
960104  Juvenile_myopathy  criteria_provided                  Benign   
960105  Juvenile_myopathy  criteria_provided  Uncertain_significance   

                            CLNVC                             MC  ORIGIN  \
0       single_nucleotide_variant    SO:0001583|missense_variant     1.0   
1       single_nucleotide_variant      SO:0001627|intron_variant     1.0   
2       single_nucleotide_variant    SO:0001583|missense_variant     1.0   
3       single_nucleotide_variant  SO:0001819|synonymous_variant     1.0   
4       single_nucleotide_variant    SO:0001583|missense_variant     1.0   
...                           ...                            ...     ...   
960101  single_nucleotide_variant                            NaN     1.0   
960102  single_nucleotide_variant                            NaN     1.0   
960103                Duplication                            NaN     1.0   
960104  single_nucleotide_variant                            NaN     1.0   
960105  single_nucleotide_variant                            NaN     1.0   

        numalt  altlen_1  is_snp  
0            1         0    True  
1            1         0    True  
2            1         0    True  
3            1         0    True  
4            1         0    True  
...        ...       ...     ...  
960101       1         0    True  
960102       1         0    True  
960103       1        15   False  
960104       1         0    True  
960105       1         0    True  

[960106 rows x 15 columns]

In [13]:
df_1.isnull().sum()

CHROM             0
POS               0
ID                0
REF               0
ALT_1           609
ALLELEID          0
CLNDN           606
CLNREVSTAT        0
CLNSIG          606
CLNVC             0
MC            16106
ORIGIN         5268
numalt            0
altlen_1          0
is_snp            0
dtype: int64

df_1 consnsits of 606 null values in CLNDN column and we can also see same numbe of null values are present in CLNSIG column,  so we will check the table and confirm if the null values are present at the same row level.

In [14]:
df_1[df_1.CLNDN.isnull()]

CHROM        POS       ID                           REF  \
5766       1    8021910   487088                             G   
5805       1    8045031   487089                             G   
6409       1   10035809   982549                             G   
8912       1   12008092   242450                             C   
15282      1   25611101   635094                             G   
...      ...        ...      ...                           ...   
953331     X  153296212   979165  CTGCTGCGCCCCTTGGGGCTGCTCTCCT   
953351     X  153296242  1172795  CTTTTCCGCCCAGGGCTCTTACAGGTCT   
953377     X  153296290   560201                           TCA   
956311     X  153762653   242750                             G   
956340     X  153763551   242797                             G   

                      ALT_1  ALLELEID CLNDN  \
5766    GGTGCTGGACGGTGTCCCT     75266   NaN   
5805                      A     22107   NaN   
6409                      A    970649   NaN   
8912                      T    215016   NaN   
15282                     T    622837   NaN   
...                     ...       ...   ...   
953331                    C    967029   NaN   
953351                    C   1161881   NaN   
953377                    T    551241   NaN   
956311                    A     38393   NaN   
956340                    C     25431   NaN   

                                      CLNREVSTAT CLNSIG  \
5766    no_interpretation_for_the_single_variant    NaN   
5805    no_interpretation_for_the_single_variant    NaN   
6409    no_interpretation_for_the_single_variant    NaN   
8912    no_interpretation_for_the_single_variant    NaN   
15282   no_interpretation_for_the_single_variant    NaN   
...                                          ...    ...   
953331  no_interpretation_for_the_single_variant    NaN   
953351  no_interpretation_for_the_single_variant    NaN   
953377  no_interpretation_for_the_single_variant    NaN   
956311  no_interpretation_for_the_single_variant    NaN   
956340  no_interpretation_for_the_single_variant    NaN   

                            CLNVC                             MC  ORIGIN  \
5766                  Duplication      SO:0001627|intron_variant     1.0   
5805    single_nucleotide_variant    SO:0001583|missense_variant     1.0   
6409    single_nucleotide_variant            SO:0001587|nonsense     1.0   
8912    single_nucleotide_variant    SO:0001583|missense_variant     1.0   
15282   single_nucleotide_variant    SO:0001583|missense_variant     0.0   
...                           ...                            ...     ...   
953331                   Deletion    SO:0001822|inframe_deletion    16.0   
953351                   Deletion    SO:0001822|inframe_deletion    16.0   
953377                   Deletion  SO:0001589|frameshift_variant     1.0   
956311  single_nucleotide_variant    SO:0001583|missense_variant     1.0   
956340  single_nucleotide_variant    SO:0001583|missense_variant     1.0   

        numalt  altlen_1  is_snp  
5766         1        18   False  
5805         1         0    True  
6409         1         0    True  
8912         1         0    True  
15282        1         0    True  
...        ...       ...     ...  
953331       1       -27   False  
953351       1       -27   False  
953377       1        -2   False  
956311       1         0    True  
956340       1         0    True  

[606 rows x 15 columns]

We will use dropna function on CLNDN to remove the null values. as we can see CLNSIG also is null on the same row level

In [15]:
df_1.dropna(subset=['CLNDN'], inplace=True)

In [16]:
df_1.isnull().sum()

CHROM             0
POS               0
ID                0
REF               0
ALT_1           589
ALLELEID          0
CLNDN             0
CLNREVSTAT        0
CLNSIG            0
CLNVC             0
MC            16092
ORIGIN         5247
numalt            0
altlen_1          0
is_snp            0
dtype: int64

After dropping all the rows from CLNDN column we can see that the rows from CLNSIG also is free of null values.

Since the percent null values in mc and ORIGIN is 1.68% and 0.54% respectively, we will drop the rows. same is the case with ALT_1, which is 0.06%

In [17]:
df_1.dropna(subset=['MC'], inplace=True)
df_1.dropna(subset=['ORIGIN'], inplace=True)

In [18]:
df_1.dropna(subset=['ALT_1'], inplace=True)

In [19]:
df_1

CHROM     POS       ID REF ALT_1  ALLELEID  \
0          1  861332  1019397   G     A   1003021   
1          1  865519  1125147   C     T   1110865   
2          1  865545  1164676   G     A   1153701   
3          1  865567  1144630   C     T   1131738   
4          1  865568   846933   G     A    824438   
...      ...     ...      ...  ..   ...       ...   
960022    MT   15873   693971   T     C    680861   
960023    MT   15882   693972   G     A    680862   
960024    MT   15884   143900   G     A    153625   
960025    MT   15884   252455   G     C    247210   
960026    MT   15885   693973   C     T    680863   

                                   CLNDN         CLNREVSTAT  \
0                           not_provided  criteria_provided   
1                           not_provided  criteria_provided   
2                           not_provided  criteria_provided   
3                           not_provided  criteria_provided   
4                           not_provided  criteria_provided   
...                                  ...                ...   
960022                    Leigh_syndrome  criteria_provided   
960023                    Leigh_syndrome  criteria_provided   
960024  Neoplasm_of_ovary|Leigh_syndrome  criteria_provided   
960025      Leigh_syndrome|not_specified  criteria_provided   
960026                    Leigh_syndrome  criteria_provided   

                        CLNSIG                      CLNVC  \
0       Uncertain_significance  single_nucleotide_variant   
1                Likely_benign  single_nucleotide_variant   
2                       Benign  single_nucleotide_variant   
3                Likely_benign  single_nucleotide_variant   
4       Uncertain_significance  single_nucleotide_variant   
...                        ...                        ...   
960022  Uncertain_significance  single_nucleotide_variant   
960023  Uncertain_significance  single_nucleotide_variant   
960024                  Benign  single_nucleotide_variant   
960025                  Benign  single_nucleotide_variant   
960026                  Benign  single_nucleotide_variant   

                                   MC  ORIGIN  numalt  altlen_1  is_snp  
0         SO:0001583|missense_variant     1.0       1         0    True  
1           SO:0001627|intron_variant     1.0       1         0    True  
2         SO:0001583|missense_variant     1.0       1         0    True  
3       SO:0001819|synonymous_variant     1.0       1         0    True  
4         SO:0001583|missense_variant     1.0       1         0    True  
...                               ...     ...     ...       ...     ...  
960022    SO:0001583|missense_variant     1.0       1         0    True  
960023            SO:0001587|nonsense     1.0       1         0    True  
960024    SO:0001583|missense_variant     1.0       1         0    True  
960025    SO:0001583|missense_variant     1.0       1         0    True  
960026    SO:0001583|missense_variant     1.0       1         0    True  

[937687 rows x 15 columns]

In [20]:
df_1.isnull().sum()

CHROM         0
POS           0
ID            0
REF           0
ALT_1         0
ALLELEID      0
CLNDN         0
CLNREVSTAT    0
CLNSIG        0
CLNVC         0
MC            0
ORIGIN        0
numalt        0
altlen_1      0
is_snp        0
dtype: int64

As per the problem given 
Classifications are divided into three categories:
1. Pathogenic or Likely pathogenic.
2. Variants of Uncertain Significance (VUS) and
3. Benign or Likely Benign.

We will find the value counts of "CLNSIG" columns and convert them to comman values like for example 'Likely_benign' will be considered as 'Benign'. 
drug_response, risk_factor, association, Affects, protective, confers_sensitivity and association_not_found values will be considered as others.
So we have 4 values present in this columns which will be easy for our classification model.

In [21]:
df_1['CLNSIG'].value_counts()

Uncertain_significance                          387233
Likely_benign                                   235151
Benign                                           97020
Pathogenic                                       88718
Conflicting_interpretations_of_pathogenicity     48453
Likely_pathogenic                                40588
Benign/Likely_benign                             19233
Pathogenic/Likely_pathogenic                      9405
not_provided                                      7592
drug_response                                     1958
other                                             1545
risk_factor                                        379
association                                        240
Affects                                            133
protective                                          30
confers_sensitivity                                  7
association_not_found                                2
Name: CLNSIG, dtype: int64

We will create a dictionary of key and values where in keys we will put the value which has to be changed and values will be the value whcih will be asigned. Next setp is to use pandas replace function and use dictionary in its arguments.

In [22]:
d ={'Likely_benign': 'Benign', 
    'Pathogenic/Likely_pathogenic':'Pathogenic',
    'Likely_pathogenic': 'Pathogenic',
    'Benign/Likely_benign':'Benign', 
    'not_provided': 'Uncertain_significance',
    'Conflicting_interpretations_of_pathogenicity' : 'Uncertain_significance',
    'drug_response':'other', 
    'risk_factor':'other', 
    'association': 'other', 
    'Affects':'other', 
    'protective' : 'other', 
    'confers_sensitivity': 'other', 
    'association_not_found' : 'other' }

In [23]:
df_1 = df_1.replace({"CLNSIG": d})

In [24]:
df_1

CHROM     POS       ID REF ALT_1  ALLELEID  \
0          1  861332  1019397   G     A   1003021   
1          1  865519  1125147   C     T   1110865   
2          1  865545  1164676   G     A   1153701   
3          1  865567  1144630   C     T   1131738   
4          1  865568   846933   G     A    824438   
...      ...     ...      ...  ..   ...       ...   
960022    MT   15873   693971   T     C    680861   
960023    MT   15882   693972   G     A    680862   
960024    MT   15884   143900   G     A    153625   
960025    MT   15884   252455   G     C    247210   
960026    MT   15885   693973   C     T    680863   

                                   CLNDN         CLNREVSTAT  \
0                           not_provided  criteria_provided   
1                           not_provided  criteria_provided   
2                           not_provided  criteria_provided   
3                           not_provided  criteria_provided   
4                           not_provided  criteria_provided   
...                                  ...                ...   
960022                    Leigh_syndrome  criteria_provided   
960023                    Leigh_syndrome  criteria_provided   
960024  Neoplasm_of_ovary|Leigh_syndrome  criteria_provided   
960025      Leigh_syndrome|not_specified  criteria_provided   
960026                    Leigh_syndrome  criteria_provided   

                        CLNSIG                      CLNVC  \
0       Uncertain_significance  single_nucleotide_variant   
1                       Benign  single_nucleotide_variant   
2                       Benign  single_nucleotide_variant   
3                       Benign  single_nucleotide_variant   
4       Uncertain_significance  single_nucleotide_variant   
...                        ...                        ...   
960022  Uncertain_significance  single_nucleotide_variant   
960023  Uncertain_significance  single_nucleotide_variant   
960024                  Benign  single_nucleotide_variant   
960025                  Benign  single_nucleotide_variant   
960026                  Benign  single_nucleotide_variant   

                                   MC  ORIGIN  numalt  altlen_1  is_snp  
0         SO:0001583|missense_variant     1.0       1         0    True  
1           SO:0001627|intron_variant     1.0       1         0    True  
2         SO:0001583|missense_variant     1.0       1         0    True  
3       SO:0001819|synonymous_variant     1.0       1         0    True  
4         SO:0001583|missense_variant     1.0       1         0    True  
...                               ...     ...     ...       ...     ...  
960022    SO:0001583|missense_variant     1.0       1         0    True  
960023            SO:0001587|nonsense     1.0       1         0    True  
960024    SO:0001583|missense_variant     1.0       1         0    True  
960025    SO:0001583|missense_variant     1.0       1         0    True  
960026    SO:0001583|missense_variant     1.0       1         0    True  

[937687 rows x 15 columns]

We will again check the value counts to ensure if our replacement has taken place successfully

In [25]:
df_1.CLNSIG.value_counts()

Uncertain_significance    443278
Benign                    351404
Pathogenic                138711
other                       4294
Name: CLNSIG, dtype: int64

Previously we have done two steps ie handling the null values and columns with unique values as one and working on CLNSIG column.

- Next we will focus on the position of each chromosome and check wha the daya tells us

* We need to get to every chromosome individually and check for for its positions. Some positions will have only one entry which some may have many entries. For the multiple entries we can see differece in the Alternate and Reference.

For our good understanding we will first filter out chromosome 1 and check its positions

In [26]:
#we will create a new dataframe which will have data of only one chromosome ie chromosome 1
chrom1_df = df_1[df_1['CHROM'] == 1]
chrom1_df

CHROM        POS       ID  REF ALT_1  ALLELEID          CLNDN  \
0         1     861332  1019397    G     A   1003021   not_provided   
1         1     865519  1125147    C     T   1110865   not_provided   
2         1     865545  1164676    G     A   1153701   not_provided   
3         1     865567  1144630    C     T   1131738   not_provided   
4         1     865568   846933    G     A    824438   not_provided   
...     ...        ...      ...  ...   ...       ...            ...   
76604     1  248616830   916385    C     T    904391   not_provided   
76605     1  248651927   403273    A     G    389409  not_specified   
76606     1  248790108   775702  CAG     C    696649   not_provided   
76607     1  249144610   788693    G     A    707276   not_provided   
76608     1  249152412   623466    C     G    612367  Severe_Myopia   

                           CLNREVSTAT                  CLNSIG  \
0                   criteria_provided  Uncertain_significance   
1                   criteria_provided                  Benign   
2                   criteria_provided                  Benign   
3                   criteria_provided                  Benign   
4                   criteria_provided  Uncertain_significance   
...                               ...                     ...   
76604               criteria_provided                  Benign   
76605               criteria_provided                  Benign   
76606               criteria_provided                  Benign   
76607               criteria_provided                  Benign   
76608  no_assertion_criteria_provided  Uncertain_significance   

                           CLNVC                             MC  ORIGIN  \
0      single_nucleotide_variant    SO:0001583|missense_variant     1.0   
1      single_nucleotide_variant      SO:0001627|intron_variant     1.0   
2      single_nucleotide_variant    SO:0001583|missense_variant     1.0   
3      single_nucleotide_variant  SO:0001819|synonymous_variant     1.0   
4      single_nucleotide_variant    SO:0001583|missense_variant     1.0   
...                          ...                            ...     ...   
76604  single_nucleotide_variant  SO:0001819|synonymous_variant     1.0   
76605  single_nucleotide_variant    SO:0001583|missense_variant     1.0   
76606                   Deletion            SO:0001587|nonsense     1.0   
76607  single_nucleotide_variant  SO:0001819|synonymous_variant     1.0   
76608  single_nucleotide_variant    SO:0001583|missense_variant     0.0   

       numalt  altlen_1  is_snp  
0           1         0    True  
1           1         0    True  
2           1         0    True  
3           1         0    True  
4           1         0    True  
...       ...       ...     ...  
76604       1         0    True  
76605       1         0    True  
76606       1        -2   False  
76607       1         0    True  
76608       1         0    True  

[74765 rows x 15 columns]

In [27]:
# we will check the values and their counts of the position
chrom1_df['POS'].value_counts()

92944314     18
55524339     11
100661987    10
20978294     10
168262522    10
             ..
6012770       1
35370851      1
43224932      1
173808690     1
220087603     1
Name: POS, Length: 70318, dtype: int64

In [28]:
chrom1_df[chrom1_df['POS'] == 92944314]

CHROM       POS       ID                    REF                  ALT_1  \
30796     1  92944314   402897                      C                    CAG   
30797     1  92944314   298179                      C                  CAGAG   
30798     1  92944314  1170883                      C                CAGAGAG   
30799     1  92944314   298180                      C              CAGAGAGAG   
30800     1  92944314   298183                      C            CAGAGAGAGAG   
30801     1  92944314   770490                      C          CAGAGAGAGAGAG   
30802     1  92944314   298181                      C        CAGAGAGAGAGAGAG   
30803     1  92944314   538141                      C      CAGAGAGAGAGAGAGAG   
30804     1  92944314   788327                      C    CAGAGAGAGAGAGAGAGAG   
30805     1  92944314  1114439                      C  CAGAGAGAGAGAGAGAGAGAG   
30806     1  92944314   800061                      C            CAGGGAGAGAG   
30807     1  92944314   298182                    CAG                      C   
30808     1  92944314   298184                  CAGAG                      C   
30809     1  92944314   298185                CAGAGAG                      C   
30810     1  92944314   470696              CAGAGAGAG                      C   
30811     1  92944314   298186            CAGAGAGAGAG                      C   
30812     1  92944314   722828          CAGAGAGAGAGAG                      C   
30813     1  92944314  1128817  CAGAGAGAGAGAGAGAGAGAG                      C   

       ALLELEID                                              CLNDN  \
30796    390535                    Severe_congenital_neutropenia_2   
30797    281299  Severe_congenital_neutropenia|Severe_congenita...   
30798   1153696                    Severe_congenital_neutropenia_2   
30799    283368  Severe_congenital_neutropenia|Severe_congenita...   
30800    281300  Severe_congenital_neutropenia|Severe_congenita...   
30801    777112                    Severe_congenital_neutropenia_2   
30802    283196  Severe_congenital_neutropenia|Severe_congenita...   
30803    516242                    Severe_congenital_neutropenia_2   
30804    778922                    Severe_congenital_neutropenia_2   
30805   1089332                    Severe_congenital_neutropenia_2   
30806    787070                    Severe_congenital_neutropenia_2   
30807    281295  Severe_congenital_neutropenia|Severe_congenita...   
30808    281937  Severe_congenital_neutropenia|Severe_congenita...   
30809    283353  Severe_congenital_neutropenia|Severe_congenita...   
30810    448256                    Severe_congenital_neutropenia_2   
30811    283195  Severe_congenital_neutropenia|Severe_congenita...   
30812    743773                    Severe_congenital_neutropenia_2   
30813   1110857                    Severe_congenital_neutropenia_2   

              CLNREVSTAT                  CLNSIG           CLNVC  \
30796  criteria_provided                  Benign  Microsatellite   
30797  criteria_provided  Uncertain_significance  Microsatellite   
30798  criteria_provided                  Benign  Microsatellite   
30799  criteria_provided  Uncertain_significance  Microsatellite   
30800  criteria_provided  Uncertain_significance  Microsatellite   
30801  criteria_provided                  Benign  Microsatellite   
30802  criteria_provided  Uncertain_significance  Microsatellite   
30803  criteria_provided                  Benign  Microsatellite   
30804  criteria_provided                  Benign  Microsatellite   
30805  criteria_provided                  Benign  Microsatellite   
30806  criteria_provided                  Benign       Insertion   
30807  criteria_provided  Uncertain_significance  Microsatellite   
30808  criteria_provided  Uncertain_significance  Microsatellite   
30809  criteria_provided                  Benign  Microsatellite   
30810  criteria_provided                  Benign  Microsatellite   
30811  criteria_provided  Uncertain_significance  Microsatellite  

We have filtered chromosome 1 position 92944314, which has 18 entries. if we observe the CLNSIG column we can find different values.  As per the given problem classifications are said to be conflicting if two of the three categories are present for a variant, such case is considered as conflicting. 

* Like wise we have to go through every position and find if CLNSIG columns have one category of many.

But initially we will create dataframes with individual chromosomes with multiple entries, since with one variant entry we cannot classify them to be conflicting or not.

In [29]:
# Creating a list of individual chromosomes
chrom_list = list(df_1.CHROM.unique())
chrom_list

[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 'X',
 'Y',
 'MT']

We need to create a function whcih gives us a dataframe for individual chromosome consisting of positions with multiple variation. Position with single variation will be dropped.

So the function works in this way:
1. The function takes two arguments 1 a dataframe and the chromosome number
2. Depending on the chromosome number the dataframe is filtered out in the next step and stored in 'x' variable.
3. in order to identify positions with multiple variation we have to find value counts of 'x', and create a new table using   reset index. (We have a table which has position number and its number of values)
4. This new table columns will be replaced as 'POS' and 'num, for our understanding  
5. We will filter out num greater than 1, since we want positions with multiple values, and create x_new table form it.
6. We will create a 'c' varable and convert the positions to numpy array and store in 'c' variable.
7. We will return a chrom dataframe(which will have all the positions which are in variable 'c') by using apply function in pandas.

In [30]:
def get_chrom(df, chrom_no):
    chrom = df[df['CHROM'] == chrom_no] 
    x = chrom['POS'].value_counts()
    x = x.reset_index()
    x.columns= ['POS', 'num']
    x_new= x.loc[x['num'] > 1]
    c = x_new['POS'].to_numpy()
    chrom['POS'] = chrom['POS'].apply(lambda x : x if x in c else None)
    chrom.dropna(subset = ['POS'], inplace=True)
    return chrom

In [31]:
# we will asign the function to c1,  check if its working properly or not
c1 = get_chrom(df_1, 1)

In [32]:
c1

CHROM          POS       ID REF ALT_1  ALLELEID  \
34        1     871143.0  1127328   G     A   1110870   
35        1     871143.0  1158587   G     T   1131742   
46        1     871215.0  1170011   C     G   1153710   
47        1     871215.0  1169806   C     T   1153711   
66        1     874456.0  1122639   G     A   1110873   
...     ...          ...      ...  ..   ...       ...   
76517     1  247599355.0    97960   A     G    103852   
76530     1  247607319.0  1096997   C     G   1067166   
76531     1  247607319.0   795624   C     T    780649   
76541     1  247607421.0  1160486   C     G   1131312   
76542     1  247607421.0   546637   C     T    537126   

                                                   CLNDN         CLNREVSTAT  \
34                                          not_provided  criteria_provided   
35                                          not_provided  criteria_provided   
46                                          not_provided  criteria_provided   
47                                          not_provided  criteria_provided   
66                                          not_provided  criteria_provided   
...                                                  ...                ...   
76517  Familial_cold_autoinflammatory_syndrome_1|not_...  criteria_provided   
76530             Cryopyrin_associated_periodic_syndrome  criteria_provided   
76531  Cryopyrin_associated_periodic_syndrome|not_pro...  criteria_provided   
76541             Cryopyrin_associated_periodic_syndrome  criteria_provided   
76542  Cryopyrin_associated_periodic_syndrome|not_spe...  criteria_provided   

           CLNSIG                      CLNVC                             MC  \
34         Benign  single_nucleotide_variant      SO:0001627|intron_variant   
35         Benign  single_nucleotide_variant      SO:0001627|intron_variant   
46         Benign  single_nucleotide_variant  SO:0001819|synonymous_variant   
47         Benign  single_nucleotide_variant  SO:0001819|synonymous_variant   
66         Benign  single_nucleotide_variant    SO:0001583|missense_variant   
...           ...                        ...                            ...   
76517  Pathogenic  single_nucleotide_variant    SO:0001583|missense_variant   
76530      Benign  single_nucleotide_variant  SO:0001819|synonymous_variant   
76531      Benign  single_nucleotide_variant  SO:0001819|synonymous_variant   
76541      Benign  single_nucleotide_variant  SO:0001819|synonymous_variant   
76542      Benign  single_nucleotide_variant  SO:0001819|synonymous_variant   

       ORIGIN  numalt  altlen_1  is_snp  
34        1.0       1         0    True  
35        1.0       1         0    True  
46        1.0       1         0    True  
47        1.0       1         0    True  
66        1.0       1         0    True  
...       ...     ...       ...     ...  
76517     1.0       1         0    True  
76530     1.0       1         0    True  
76531     1.0       1         0    True  
76541     1.0       1         0    True  
76542     1.0       1         0    True  

[8402 rows x 15 columns]

our function worked fine since we can see in dataframe above.

We will also create a function which will add a new column "CONFLICTING" which will have 0 and 1, 0 suggesting there is only one category in 'CLNSIG' column and 1 if there are more categories.

The function will be as follows:
1. The function will take one argument ie a dataframe.
1. For the dataframe we created it will add a column "CONFLICTING" with 0 value in it.
2. We will create a list of Position present in the dataframe, using set function we will remove the duplicates
3. We will loop over the position list we created, and filter each and every position individually and check for CLNSIG category
4. We will use set function and turn it into len, and if there are multiple categories the length will be more than 1.
5. If it is more than 1 then the CONFLICTING value will be converted to 1.


In [33]:
def conflicting(c):
    c['CONFLICTING'] = 0
    pos_list = list(set(c.POS))
    for i in pos_list: 
        if len(set(c[c['POS'] == i]['CLNSIG'])) != 1:
            c.loc[c.POS == i, "CONFLICTING"] = 1      
        else:
            continue
    return c

Our next step is to use both the functions we have created
* First function 'get_chrom' is to get all the chromosomes dataframes with multiple variation entries
* The second function is to assign if the categories present in every dataframe hase more than one values or not.

We will use python exec function to create different variable names, and we will loop throuch the chromosome list we creaated to get the names. We will create individual dataframe for each chromosome.

In [34]:
for i in chrom_list:
    exec(f'c_{i} = get_chrom(df_1, i)')
    print(f'Chrom {i} created ...')
print('Done')

Chrom 1 created ...
Chrom 2 created ...
Chrom 3 created ...
Chrom 4 created ...
Chrom 5 created ...
Chrom 6 created ...
Chrom 7 created ...
Chrom 8 created ...
Chrom 9 created ...
Chrom 10 created ...
Chrom 11 created ...
Chrom 12 created ...
Chrom 13 created ...
Chrom 14 created ...
Chrom 15 created ...
Chrom 16 created ...
Chrom 17 created ...
Chrom 18 created ...
Chrom 19 created ...
Chrom 20 created ...
Chrom 21 created ...
Chrom 22 created ...
Chrom 23 created ...
Chrom X created ...
Chrom Y created ...
Chrom MT created ...
Done


In [47]:
new_chrom_list = [c_1, c_2, c_3, c_4, c_5, c_6, c_7, c_8, c_9, c_10,
              c_11, c_12, c_13, c_14, c_15, c_16, c_17, c_18, c_19, c_20,
              c_21, c_22, c_23, c_X, c_Y, c_MT ]

new_chrom_str = ['c_1', 'c_2', 'c_3', 'c_4', 'c_5', 'c_6', 'c_7', 'c_8', 'c_9', 'c_10',
              'c_11', 'c_12', 'c_13', 'c_14', 'c_15', 'c_16', 'c_17', 'c_18', 'c_19', 'c_20',
              'c_21', 'c_22', 'c_23', 'c_X', 'c_Y', 'c_MT' ]

We will use python zip function to apply the values from the list we have created above ie new_chrom_list(which has all the name of dataframes) and new_chrom_str(which has all the names of dataframe stroed as strings). We will use for loop and create individual dataframe for each chromosome using conflicting function.

In [48]:
for i,j in zip(new_chrom_str, new_chrom_list):
    exec(f'{i}_final = conflicting(j)')
    print(f'{i} created ...')
print('Done')

c_1 created ...
c_2 created ...
c_3 created ...
c_4 created ...
c_5 created ...
c_6 created ...
c_7 created ...
c_8 created ...
c_9 created ...
c_10 created ...
c_11 created ...
c_12 created ...
c_13 created ...
c_14 created ...
c_15 created ...
c_16 created ...
c_17 created ...
c_18 created ...
c_19 created ...
c_20 created ...
c_21 created ...
c_22 created ...
c_23 created ...
c_X created ...
c_Y created ...
c_MT created ...
Done


Our final step is to concatinate all the dataframe we have created using concat function in pandas.

In [49]:
all_df =[c_1_final, c_2_final, c_3_final, c_4_final, c_5_final, c_6_final, c_7_final, c_8_final, c_9_final, c_10_final,
         c_11_final, c_12_final, c_13_final, c_14_final, c_15_final, c_16_final, c_17_final, c_18_final, c_19_final, c_20_final, 
         c_21_final, c_22_final, c_23_final, c_X_final,c_Y_final,c_MT_final]
final_df = pd.concat(all_df)

In [50]:
final_df

CHROM       POS       ID REF ALT_1  ALLELEID  \
34         1  871143.0  1127328   G     A   1110870   
35         1  871143.0  1158587   G     T   1131742   
46         1  871215.0  1170011   C     G   1153710   
47         1  871215.0  1169806   C     T   1153711   
66         1  874456.0  1122639   G     A   1110873   
...      ...       ...      ...  ..   ...       ...   
960019    MT   15852.0   693968   T     G    680858   
960020    MT   15860.0   693970   A     G    680859   
960021    MT   15860.0   693969   A     T    680860   
960024    MT   15884.0   143900   G     A    153625   
960025    MT   15884.0   252455   G     C    247210   

                                   CLNDN         CLNREVSTAT  \
34                          not_provided  criteria_provided   
35                          not_provided  criteria_provided   
46                          not_provided  criteria_provided   
47                          not_provided  criteria_provided   
66                          not_provided  criteria_provided   
...                                  ...                ...   
960019                    Leigh_syndrome  criteria_provided   
960020                    Leigh_syndrome  criteria_provided   
960021                    Leigh_syndrome  criteria_provided   
960024  Neoplasm_of_ovary|Leigh_syndrome  criteria_provided   
960025      Leigh_syndrome|not_specified  criteria_provided   

                        CLNSIG                      CLNVC  \
34                      Benign  single_nucleotide_variant   
35                      Benign  single_nucleotide_variant   
46                      Benign  single_nucleotide_variant   
47                      Benign  single_nucleotide_variant   
66                      Benign  single_nucleotide_variant   
...                        ...                        ...   
960019  Uncertain_significance  single_nucleotide_variant   
960020                  Benign  single_nucleotide_variant   
960021                  Benign  single_nucleotide_variant   
960024                  Benign  single_nucleotide_variant   
960025                  Benign  single_nucleotide_variant   

                                   MC  ORIGIN  numalt  altlen_1  is_snp  \
34          SO:0001627|intron_variant     1.0       1         0    True   
35          SO:0001627|intron_variant     1.0       1         0    True   
46      SO:0001819|synonymous_variant     1.0       1         0    True   
47      SO:0001819|synonymous_variant     1.0       1         0    True   
66        SO:0001583|missense_variant     1.0       1         0    True   
...                               ...     ...     ...       ...     ...   
960019    SO:0001583|missense_variant     1.0       1         0    True   
960020    SO:0001583|missense_variant     1.0       1         0    True   
960021    SO:0001583|missense_variant     1.0       1         0    True   
960024    SO:0001583|missense_variant     1.0       1         0    True   
960025    SO:0001583|missense_variant     1.0       1         0    True   

        CONFLICTING  
34                0  
35                0  
46                0  
47                0  
66                1  
...             ...  
960019            1  
960020            0  
960021            0  
960024            0  
960025            0  

[156822 rows x 16 columns]

We have our final dataset where we can perform statistical analysis and create classification model effectively.

In [53]:
final_df.to_csv('final_df.csv',index=False)# converting final_df to csv file for further process